npeps/npep-187-ports-and-protocols.md

```diff
@@ -0,0 +1,408 @@
+For now, it can be expressed as a port range that covers all ports (1-65535).
+
+In the `protocols` filter design we decided to start from the YAML that we want (i.e. YAML that is easy to write and understand).
+We want to make popular filters (TCP and UDN ports) **easy** and less popular filters (like ICMP and named ports) **possible**.
+There are multiple options to support this:
+
+### Option 1: Protocols list with protocol discriminator
+
+```yaml
+protocols:
+  - protocol: TCP
+    port: 
+      number: 8080
+  - protocol: UDP
+    port:
+      range: 
+        start: 8080
+        end: 9090 
+  - namedPort: http
+    # protocol: "" # must be empty string: defaulted / omitempty
+```
+
+with the validation like so:
+- empty `protocols` list is not allowed
+- at least 1 field for each `protocol` element must be set
+- if `namedPort` is set, `protocol` must be unset
+- if `protocol` is TCP/UDP/SCTP, `port` must be set
+- exactly one of `number` or `range` must be set in `port`
+
+### Option2: Join by protocol
+
+```yaml
+protocols:
+  tcp:
+    - port:
+        number: 8080
+    - port:
+        range: 
+          start: 8080
+          end: 9090
+  udp:
+    - port:
+        number: 8080
+    - port:
+        number: 9090
+  namedPort:
+    - http
+    - monitoring
+  icmp:
+    - type: 7
+      code: 3
+```
+
+In this case it makes more sense to also flatten the `port` match, because for TCP/UDP/SCTP protocols, port is the only option and is always required:
+
+```yaml
+protocols:
+  tcp:
+    - port: 8080
+    - portRange: 
+        start: 8080
+        end: 9090
+  udp:
+    - port: 8080
+    - port: 9090
+  namedPort:
+    - http
+    - monitoring
+  icmp:
+    - type: 7
+      code: 3
+```
+
+with the validation like so:
+- empty `protocols` list is not allowed
+- at least 1 field of `protocol` must be set
+- each protocol list must not be empty
+- exactly one of `port` or `portRange` must be set for TCP/UDP/SCTP elements
+
+### Option3: Support older clients
+
+Supporting older clients that may not know about the newer API fields is an important consideration for APIs where
+evolution is highly likely. The idea is to make it possible for older clients to understand/report that a
+given CNP has fields they don't understand and "fail closed".
+
+Suppose one day we add GRE support. The API understands an object with gre specified but we have an 
+older client which does not. The client will deserialize the gre field and discard it. 
+One approach to defending that is to elevate the list, like:
+
+```go
+type Protocols struct {
+// exactly one should be set, if you see this empty it's something you don't know how to handle
+    TCP *TCPProtocol
+    UDP *UDPProtocol
+    GRE *GREProtocol
+}
+```
+
+Then rather than the parent having a single `Protocols Protocols` it would have a list: `Protocols []Protocols`. Rendering as YAML like:
+
+```yaml
+protocols
+- tcp:
+    port: 53
+- udp:
+    port: 53
+- gre:
+    somethingHere: true
+```
```

Comment:
This has the advantage that "don't restrict on protocol" can be represented by the empty list without defining a sigil.